File: CPP/statistics_calculator.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>
#include <cmath> // For isnan()
#include <cfloat> // For NAN on some older systems

// Function to calculate the mean
double calculateMean(const std::vector<int>& numbers) {
    if (numbers.empty()) {
        return NAN;
    }
    double sum = 0;
    for (size_t i = 0; i < numbers.size(); ++i) { // Use index-based loop
        sum += numbers[i];
    }
    return sum / numbers.size();
}

// Function to calculate the median
double calculateMedian(const std::vector<int>& numbers) {
    if (numbers.empty()) {
        return NAN;
    }
    std::vector<int> sortedNumbers = numbers; // Copy the vector
    std::sort(sortedNumbers.begin(), sortedNumbers.end());
    size_t size = sortedNumbers.size();
    if (size % 2 == 0) {
        return (sortedNumbers[size / 2 - 1] + sortedNumbers[size / 2]) / 2.0;
    } else {
        return sortedNumbers[size / 2];
    }
}

// Function to calculate the mode(s)
std::vector<int> calculateMode(const std::vector<int>& numbers) {
    if (numbers.empty()) {
        return std::vector<int>(); // Use constructor for empty vector
    }
    std::vector<int> sortedNumbers = numbers;
    std::sort(sortedNumbers.begin(), sortedNumbers.end());
    size_t size = sortedNumbers.size();

    int maxCount = 0;
    int currentCount = 1;
    std::vector<int> modes;

    for (size_t i = 1; i < size; ++i) { // Use index-based loop
        if (sortedNumbers[i] == sortedNumbers[i - 1]) {
            currentCount++;
        } else {
            if (currentCount > maxCount) {
                maxCount = currentCount;
                modes.clear();
                modes.push_back(sortedNumbers[i - 1]);
            } else if (currentCount == maxCount) {
                modes.push_back(sortedNumbers[i - 1]);
            }
            currentCount = 1;
        }
    }
     if (currentCount > maxCount) {
        modes.clear();
        modes.push_back(sortedNumbers[size-1]);
    }
    else if(currentCount == maxCount){
         modes.push_back(sortedNumbers[size-1]);
    }
    return modes;
}

int main() {
    std::cout << "Welcome to the Statistics Calculator! Enter a list of integers (with spaces) or type 'done': " << std::endl;

    std::string inputLine;
    std::vector<int> numbers;

    while (std::getline(std::cin, inputLine)) {
        if (inputLine == "done") {
            break;
        }

        std::stringstream ss(inputLine);
        int number;
        bool validInput = true;
        while (ss >> number) {
            numbers.push_back(number);
        }
        if (numbers.empty() && inputLine.empty())
        {
             std::cout << "No integers were entered. Please provide a list of integers or 'done': " << std::endl;
             continue;

        }

        if (ss.fail() && !ss.eof()) {
            std::cout << "Invalid input. Please enter integers separated by spaces, or 'done': " << std::endl;
            numbers.clear(); // Clear the vector for the next input
            continue; // Go back to the beginning of the while loop
        }
        break; // Exit the loop if input is valid

    }
    if (numbers.empty())
    {
        std::cout << "No numbers entered. Exiting." << std::endl;
        return 0;
    }

    double mean = calculateMean(numbers);
    if (std::isnan(mean)) {
        std::cout << "Mean: No data provided" << std::endl;
    } else {
        std::cout << "Mean: " << mean << std::endl;
    }

    double median = calculateMedian(numbers);
    if (std::isnan(median)) {
        std::cout << "Median: No data provided" << std::endl;
    } else {
        std::cout << "Median: " << median << std::endl;
    }

    std::vector<int> modes = calculateMode(numbers);
    if (modes.empty()) {
        std::cout << "Mode: No mode found" << std::endl;
    } else {
        std::cout << "Mode(s): ";
        for (size_t i = 0; i < modes.size(); ++i) { // Use index based loop
            std::cout << modes[i] << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}

File: Python/statistics_calculator.py
import math

class StatisticsCalculator:
    
    # A class for calculating basic statistics (mean, median, mode) of a list of numbers.
    
    def __init__(self, data=None):
        """
        Initializes the StatisticsCalculator with a list of data.

        Args:
            data: A list of numbers (integers or floats).  Defaults to an empty list.
        """
        self.data = list(data) if data else []  # Ensure it's a list, handle None

    def calculate_mean(self):
        """
        Calculates the mean (average) of the data.

        Returns:
            The mean of the data, or float('nan') if the data list is empty.
        """
        if not self.data:
            return float('nan')
        return sum(self.data) / len(self.data)

    def calculate_median(self):
        """
        Calculates the median of the data.

        Returns:
            The median of the data, or float('nan') if the data list is empty.
        """
        if not self.data:
            return float('nan')
        sorted_data = sorted(self.data)
        n = len(sorted_data)
        if n % 2 == 0:
            # Even number of elements, average the two middle ones
            mid1 = sorted_data[n // 2 - 1]
            mid2 = sorted_data[n // 2]
            return (mid1 + mid2) / 2
        else:
            # Odd number of elements, take the middle one
            return sorted_data[n // 2]

    def calculate_mode(self):
        """
        Calculates the mode(s) of the data.

        Returns:
            A list of the mode(s) of the data. Returns an empty list
            if the data list is empty, or if there is no unique mode.
        """
        if not self.data:
            return []

        counts = {}
        for number in self.data:
            counts[number] = counts.get(number, 0) + 1

        max_count = 0
        modes = []
        for number, count in counts.items():
            if count > max_count:
                modes = [number]
                max_count = count
            elif count == max_count:
                modes.append(number)
        return modes

def get_integer_list_input():
    """
    Gets a list of integers from the user. Handles input validation and empty input.

    Returns:
        A list of integers entered by the user.  Returns an empty list
        if no valid integers are entered.
    """
    while True:
        input_str = input("Enter a list of integers (with spaces) or type 'done': ")
        if input_str.lower() == 'done':
            return []  # Return an empty list to signal no input

        try:
            numbers = [int(x) for x in input_str.split()]
            if not numbers:
                print("No integers were entered. Please provide a list of integers or 'done'.")
                continue #restart the loop
            return numbers
        except ValueError:
            print("Invalid input. Please enter integers separated by spaces, or 'done'.")
            # No return here, loop back to the beginning of the while loop